src/os/bluestore/BlueStore.cc

```diff
@@ -19175,73 +19356,24 @@ int cmpfunc (const void * a, const void * b)
-{
-  uint64_t allocation_size = std::min((req_extent_count) * sizeof(extent_t), memory_target / 3);
-  uint64_t extent_count    = allocation_size/sizeof(extent_t);
-  dout(5) << "req_extent_count=" << req_extent_count << ", granted extent_count="<< extent_count << dendl;
-
-  unique_ptr<extent_t[]> arr1;
-  unique_ptr<extent_t[]> arr2;
-  try {
-    arr1 = make_unique<extent_t[]>(extent_count);
-    arr2 = make_unique<extent_t[]>(extent_count);
-  } catch (std::bad_alloc&) {
-    derr << "****Failed dynamic allocation, extent_count=" << extent_count << dendl;
-    return -1;
-  }
-
-  // copy the extents from the allocators into simple array and then compare them
-  uint64_t size1 = 0, size2 = 0;
-  uint64_t idx1  = 0, idx2  = 0;
-  auto iterated_mapper1 = [&](uint64_t offset, uint64_t length) {
-    size1 += length;
-    if (idx1 < extent_count) {
-      arr1[idx1++] = {offset, length};
-    }
-    else if (idx1 == extent_count) {
-      derr << "(2)compare_allocators:: spillover"  << dendl;
-      idx1 ++;
-    }
-
-  };
-
-  auto iterated_mapper2 = [&](uint64_t offset, uint64_t length) {
-    size2 += length;
-    if (idx2 < extent_count) {
-      arr2[idx2++] = {offset, length};
-    }
-    else if (idx2 == extent_count) {
-      derr << "(2)compare_allocators:: spillover"  << dendl;
-      idx2 ++;
-    }
-  };
-
-  alloc1->foreach(iterated_mapper1);
-  alloc2->foreach(iterated_mapper2);
-
-  qsort(arr1.get(), std::min(idx1, extent_count), sizeof(extent_t), cmpfunc);
-  qsort(arr2.get(), std::min(idx2, extent_count), sizeof(extent_t), cmpfunc);
-
-  if (idx1 == idx2) {
-    idx1 = idx2 = std::min(idx1, extent_count);
-    if (memcmp(arr1.get(), arr2.get(), sizeof(extent_t) * idx2) == 0) {
-      return 0;
-    }
-    derr << "Failed memcmp(arr1, arr2, sizeof(extent_t)*idx2)"  << dendl;
-    for (uint64_t i = 0; i < idx1; i++) {
-      if (memcmp(arr1.get()+i, arr2.get()+i, sizeof(extent_t)) != 0) {
-	derr << "!!!![" << i << "] arr1::<" << arr1[i].offset << "," << arr1[i].length << ">" << dendl;
-	derr << "!!!![" << i << "] arr2::<" << arr2[i].offset << "," << arr2[i].length << ">" << dendl;
-	return -1;
-      }
-    }
-    return 0;
-  } else {
-    derr << "mismatch:: idx1=" << idx1 << " idx2=" << idx2 << dendl;
-    return -1;
-  }
+int BlueStore::compare_allocators(Allocator* alloc1, Allocator* alloc2)
+{
+  uint64_t capacity = std::max(alloc1->get_capacity(), alloc2->get_capacity());
+  uint64_t alloc_size =
+    std::min(alloc1->get_block_size(), alloc2->get_block_size());
+  SimpleBitmap b1(cct, capacity / alloc_size);
+  SimpleBitmap b2(cct, capacity / alloc_size);
+  uint64_t order = std::countr_zero(alloc_size);
+
+  alloc1->foreach(
+    [&](uint64_t offset, uint64_t length) {
+      b1.set(offset >> order, length >> order);
+  });
+  alloc2->foreach(
+    [&](uint64_t offset, uint64_t length) {
+      b2.set(offset >> order, length >> order);
+  });
+  return 0;
```

Comment:
Missing actual compare between b1 and b2.

Reply:
good catch